mlir/include/mlir/Dialect/Linalg/TransformOps/LinalgTransformOps.td

```diff
@@ -1317,25 +1317,21 @@ def ConvertToLoopsOp : Op<Transform_Dialect, "structured.convert_to_loops",
 def DecomposeInterfaceOp : Op<Transform_Dialect, "structured.decompose_interface",
     [FunctionalStyleTransformOpTrait,
      MemoryEffectsOpInterface,
-     TransformOpInterface,
-     TransformEachOpTrait,
+     DeclareOpInterfaceMethods<TransformOpInterface>,
      ReportTrackingListenerFailuresOpTrait]> {
   let description = [{
-    TODO
+    Decomposes high-level named ops into a sequence of non-aggregate named ops
+    via `AggregatedOpInterface`.
+
+    The operation ignores non-decomposable ops. The return handles point to
+    a sequence of named ops produced by the decomposition.
   }];
 
   let arguments = (ins TransformHandleTypeInterface:$target);
-  let results = (outs TransformHandleTypeInterface:$transformed);
+  let results = (outs Variadic<TransformHandleTypeInterface>:$transformed);
```

Comment:
I don't think this is right.

`Variadic<..>` allows instances of an op to have different number of return values, but for each instance the number of result values still needs to be statically known, i.e. there is always a fixed number of results in the IR.

What you are doing in the implementation of `apply()` in `LinalgTransformOps.cpp` is pushing a result for each op in the payload that was associated to `target`. Note that the number of ops in this payload is only known at runtime. Hence it better be that the number of ops associated to `target` equals the number of results that the instance of `transform.structured.decompose_interface` expects, otherwise there will be a runtime error.

This issue also crops up in another situation: you optionally do not push a result if an op associated to `target` does not implement the required interface, i.e. you "ignore" it. Because a user is statically encoding the number of expected results, a user would always need to know upfront how many of the ops passed via `target` do or do not implement the interface. Otherwise the number of results would be wrong and there would be a runtime error.

Hence this scheme does not work in general. It would mean always needing to know statically how many decomposable ops are going to be associated to an instance of `decompose_interface`'s `target` argument at runtime.

Maybe @ftynse has a suggestion for still allowing the `decompose_interface` TransformOp to be "vectorized," but at the moment I don't see a sensible interface for it.

I think the interface that does work is to insist that `target` payload's size is at most one (there a number of TransformOps that insist on this). In that case you just have one return handle (so no `Variadic<..>`) which either carries all the decomposed ops or is empty in case 1) `target` was empty or 2) the associated op didn't implement the decompose interface.

Reply:
Thanks for the suggestion, Rolf! Yes, this is awkward. Here's an example I wrote while testing the thing, it highlights the usage problem:
```mlir
func.func @softmax(%arg0: tensor<2x16x32xf32>, %dst: tensor<2x16x32xf32>) -> tensor<2x16x32xf32> {
  %out = tensor.empty() : tensor<2x16x32xf32>
  %1 = linalg.softmax dimension(2) ins(%arg0 : tensor<2x16x32xf32>) outs(%out: tensor<2x16x32xf32>) -> tensor<2x16x32xf32>
  %2 = linalg.softmax dimension(1) ins(%1: tensor<2x16x32xf32>) outs(%dst: tensor<2x16x32xf32>) -> tensor<2x16x32xf32>
  return %2 : tensor<2x16x32xf32>
}

module attributes {transform.with_named_sequence} {
  transform.named_sequence @__transform_main(%arg1: !transform.any_op {transform.readonly}) {
    %2 = transform.structured.match ops{["linalg.softmax"]} in %arg1 : (!transform.any_op) -> !transform.any_op
    %3, %4 = transform.structured.decompose_interface %2 : (!transform.any_op) -> (!transform.any_op, !transform.any_op)
    %5 = transform.structured.generalize %3: (!transform.any_op) -> !transform.any_op
    %6 = transform.structured.generalize %4: (!transform.any_op) -> !transform.any_op

    transform.yield
  }
}
```
The other option I considered was to assign all the resulting ops to the single value the transform op returns. That had the downside of not knowing which new payload corresponds to which target, so I went this route.

Anyway, the purpose of the change was to introduce the pass. Transform op change is just a byproduct (and it seems the op was created for demonstration in the first place? I found no usage scenarios of it anywhere). I don't have an opinion on how the interface should look like. I'll just list some properties to consider:
* Multiple ops in the resulting decomposition (i.e., can't go through `applyToOne`)
* Should skip ops that don't implement the interface
* Would be nice to apply to a sequence (?)

If there are no other strong opinions I'll follow the suggestion.

Reply:
> The other option I considered was to assign all the resulting ops to the single value the transform op returns. That had the downside of not knowing which new payload corresponds to which target

I think this is also a valid option, namely keep the op vectorized but it's up to the user to deal with the information loss of which decomposed ops belonged to which op in `target`'s payload. For example, if you know you want to do the same thing to all decomposed ops anyway, e.g. generalize them, this would be more compact than needing use a `transform.foreach` and stuffing the `decompose_interface` op in it's region. In case the user ensures there's only one op associated to `target`, e.g. by using `transform.foreach`, then they regain the unambiguous semantics of my suggestion.

You could even make the "accumulation" behaviour opt-in by requiring the user to use a (unit-)attribute or keyword. Whether this is a pattern that is desirable for the transform dialect is I think best addressed by @ftynse.

Reply:
In any case, the above is basically just bikeshedding. If you want something that can be merged quick, you could go with my non-vectorized suggestion, which is basically an established pattern in the Transform Dialect.

The vectorized behaviour could always be added in a separate PR.

Reply:
I went with binding all the payload with a single output value (reverting the op change). This was easiest, it doesn't modify the original interface, and has less verbose usage:

```mlir
func.func @softmax(%arg0: tensor<2x16x32xf32>, %dst: tensor<2x16x32xf32>) -> tensor<2x16x32xf32> {
  %out = tensor.empty() : tensor<2x16x32xf32>
  %1 = linalg.softmax dimension(2) ins(%arg0 : tensor<2x16x32xf32>) outs(%out: tensor<2x16x32xf32>) -> tensor<2x16x32xf32>
  %2 = linalg.softmax dimension(1) ins(%1: tensor<2x16x32xf32>) outs(%dst: tensor<2x16x32xf32>) -> tensor<2x16x32xf32>
  return %2 : tensor<2x16x32xf32>
}

module attributes {transform.with_named_sequence} {
  transform.named_sequence @__transform_main(%arg1: !transform.any_op {transform.readonly}) {
    %2 = transform.structured.match ops{["linalg.softmax"]} in %arg1 : (!transform.any_op) -> !transform.any_op
    %3 = transform.structured.decompose_interface %2 : (!transform.any_op) -> (!transform.any_op)
    %5 = transform.structured.generalize %3: (!transform.any_op) -> !transform.any_op

    transform.yield
  }
}
```

Reply:
Fine by me. 👍